File: ts/128/21.ts
import { ListNode } from '../class'

// 迭代
export function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  const dummy = new ListNode(-1)
  let cur = dummy
  while (list1 || list2) {
    if (!list1) {
      cur.next = list2
      break
    }
    if (!list2) {
      cur.next = list1
      break
    }
    if (list1.val < list2.val) {
      cur.next = list1
      list1 = list1.next
    } else {
      cur.next = list2
      list2 = list2.next
    }
    cur = cur.next
  }
  return dummy.next
};

File: c/128/21.c
#include <stdio.h>
#include <stdlib.h>
#include "../structs.h"


struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2){
  struct ListNode* dummy = malloc(sizeof(struct ListNode));
  dummy->next = NULL;
  struct ListNode* ptr = dummy;
  while (list1 && list2) {
    if (list1->val < list2->val) {
      ptr->next = list1;
      list1 = list1->next;
    } else {
      ptr->next = list2;
      list2 = list2->next;
    }
    ptr = ptr->next;
  }
  if (list1 == NULL) ptr->next = list2;
  else ptr->next = list1;
  return dummy->next;
}